app/models/transfer_request.rb

```diff
@@ -170,17 +188,45 @@ def suggested_transition_to(target)
   def perform_transfer_of_contents
     return if asset.failed? || asset.cancelled?
 
-    target_asset.aliquots << asset.aliquots.map do |aliquot|
-      aliquot.dup(aliquot_attributes(aliquot))
-    end
+    target_asset.aliquots << aliquots_for_transfer
   rescue ActiveRecord::RecordNotUnique => e
     # We'll specifically handle tag clashes here so that we can produce more informative messages
     raise e unless /aliquot_tags_and_tag2s_are_unique_within_receptacle/.match?(e.message)
 
     errors.add(:asset, "contains aliquots which can't be transferred due to tag clash")
+
     raise Aliquot::TagClash, self
   end
 
+  # If merge_equivalent_aliquots is false, or unset we do not detect
+  # equivalent aliquots (ie. those with the same sample/tags/primer_panels etc)
+  # before performing a transfer.
+  # If merge_equivalent_aliquots is true, or we detect equivalent aliquots
+  # and do not attempt to transfer them. Essentially this merges the two aliquots
+  # together. It will NOT be possible to distinguish between them.
+  # This is was added to support the Heron (Covid-19) sequencing pipeline,
+  # where two plates were subject to separate PCR processes, before being
+  # merged together again.
+  def aliquots_for_transfer
+    if merge_equivalent_aliquots
+      duplicates_of_distinct_source_aliquots_only
+    else
+      duplicates_of_all_source_aliquots
+    end
+  end
+
+  def duplicates_of_all_source_aliquots
+    asset.aliquots.map do |aliquot|
+      aliquot.dup(aliquot_attributes(aliquot))
+    end
+  end
+
+  def duplicates_of_distinct_source_aliquots_only
+    duplicates_of_all_source_aliquots.reject do |candidate_aliquot|
+      target_asset.aliquots.any? { |existing_aliquot| existing_aliquot.equivalent?(candidate_aliquot) }
+    end
+  end
```

Comment:
Little bit confused about what's happening here. Will this `perform_transfer_of_contents` method run once for PCR plate 1, at which point all the aliquots will be transferred, and then a second time for PCR plate 2, at which point none of the aliquots will be transferred, because it's checking against `target_asset.aliquots`?

Reply:
Limber creates a TransferRequestCollection in SS which contains all transfer request definitions coming from both PCR1 and PCR2; for every request we run perform_transfer_of_contents so in this list of transfer request creations, elements from PCR1 and PCR2 will go to the same destination well, the first one will create the request without problem, the second will need to check if it can happen by comparing with the first.